hdf5: Fix compiler test

pjh40 opened the pull request:
Use `self.spec.satisfies` on compiler to determine if a flag should be
applied or not.  This approach avoids issues with the strings `gcc`
or `clang` appearing in the full path to the compiler executables, as
happens with spack-installed compilers (e.g. `nvhpc%gcc`).  Effectively,
this restores PR #24451, although uses a slightly different test. @skosukhin

Commit:
Fix compiler test

Use `self.spec.satisfies` on compiler to determine if a flag should be
applied or not.  This approach avoids issues with the strings `gcc`
or `clang` appearing in the full path to the compiler executables, as
happens with spack-installed compilers (e.g. `nvhpc%gcc`).

brtnfld reviewed (dismissed):
LGTM

skosukhin commented:
@pjh40 if I am not mistaken, `"clang" in self.compiler.cc` is supposed to be `True` not only for `%clang` but also for `%aocc`, `%apple-clang`, `%arm` and maybe some others. I don't know, however, whether all of them really need the flag.

pjh40 commented:
> @pjh40 if I am not mistaken, `"clang" in self.compiler.cc` is supposed to be `True` not only for `%clang` but also for `%aocc`, `%apple-clang`, `%arm` and maybe some others. I don't know, however, whether all of them really need the flag.

Hmm... is there a way to test for `clang` in the last position of the path?  The problem shows up when `gcc` is found in the middle of the `self.compiler.cc` path, like
```
spack/opt/spack/linux-rhel7-power9le/gcc-9.3.0/nvhpc-21.5-iltp4r26jmjimrxfg47pqu2an27akxzb/Linux_ppc64le/21.5/compilers/bin/nvcc
```

Commit:
Limit compiler name search to last path component

@skosukhin pointed out that the cflag modification should happen for any
clang or gcc compiler, regardless of what compiler spec provides them.
This commit reverts to searching for a compiler name containing "gcc"
or "clang", but limits the search to the last path component, which
avoids matching spack-installed compilers built with gcc (e.g.
`nvhpc%gcc`), which will have "gcc" in the compiler path.

Commit:
Use `os.path` rather than `pathlib`

sethrj commented on var/spack/repos/builtin/packages/hdf5/package.py:
```diff
@@ -187,7 +187,8 @@ def flag_handler(self, name, flags):
         if name == "cflags":
-            if "clang" in self.compiler.cc or "gcc" in self.compiler.cc:
+            cc_name = os.path.basename(self.compiler.cc)
+            if "clang" in cc_name or "gcc" in cc_name:
```
I was going to suggest testing for `isinstance(self.compiler, spack.compilers.clang.Clang)` but apparently Aocc doesn't follow this convention. This is kind of a hack but it looks likely enough to work among the alternatives...

pjh40 replied:
I agree it is pretty hacky looking, and I'm not even sure that it is necessary to jump thru hoops like this to disable warnings. However, the way it stands is completely broken for spack-installed compilers, so I was just trying to get something in that wouldn't be reverted again.

pjh40 commented:
Trying to get a ci-pipeline pass

sethrj commented:
Closing and reopening to start gitlab CI...

pjh40 commented:
Is there a way to just restart the ci/gitlab-ci pipeline?  It is dying with "runner failures".

sethrj commented:
@pjh40 Apparently closing and reopening has no effect on the runner-ci job. But fortunately @tgamblin just enabled us to judiciously "ignore" the gitlab-ci failures -- which I will do for this case.
